add __hash__ to FunctionSchema

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #90735
* #90734
* #90733
* #90732
* #90731
* __->__ #90730

This PR adds __hash__ to FunctionSchema pybind binding, so that
it could be used for things like dict indexing

wanchaol committed:
add __hash__ to FunctionSchema

This PR adds __hash__ to FunctionSchema pybind binding, so that
it could be used for things like dict indexing

[ghstack-poisoned]

ezyang commented on torch/csrc/jit/python/init.cpp:
```diff
@@ -1700,6 +1700,13 @@ void initJITBindings(PyObject* module) {
+          [](const FunctionSchema& self) {
+            std::stringstream ss;
+            ss << self;
+            return std::hash<std::string>()(ss.str());
```
This is a very inefficient way to go about hashing. More conventional way to do hashing is to inspect the equality function, hash each component and combine the hashes

wanchaol replied:
Thanks for the guide! Just updated the PR to use individual component hashing and combine them.

wanchaol committed:
Update on "add __hash__ to FunctionSchema"


This PR adds __hash__ to FunctionSchema pybind binding, so that
it could be used for things like dict indexing

[ghstack-poisoned]

ezyang commented on aten/src/ATen/core/function_schema.h:
```diff
@@ -639,6 +639,40 @@ template<>
+    size_t operator()(const c10::Argument& arg) const
+    {
+      // We don't need to hash default value and alias_info
+      // because they are not useful in distinguishing schema.
```
The hash algorithm should respect equality?

ezyang replied:
Like, if you have FunctionSchemas with differing default values hash to the same thing, then why did you just hash on the name of the schema only. Would be a lot simpler!

wanchaol replied:
I see, yeah so far I don't think we have two schemas with exact same arguments and only differs in default values, and couldn't reason about having such ones in the future, that was the primary reason I chose to not hash default value. But if we want to be more respectful to equality, I can surely add the default value hash. I think we still want to leave out the alias_info as it contains a ton of information which is difficult to hash and not really useful in distinguishing schemas. 

I'll update the PR to include default value hashing.

wanchaol committed:
Update on "add __hash__ to FunctionSchema"


This PR adds __hash__ to FunctionSchema pybind binding, so that
it could be used for things like dict indexing

[ghstack-poisoned]

wanchaol committed:
Update on "add __hash__ to FunctionSchema"


This PR adds __hash__ to FunctionSchema pybind binding, so that
it could be used for things like dict indexing

[ghstack-poisoned]

ezyang commented on aten/src/ATen/core/function_schema.h:
```diff
@@ -639,6 +639,44 @@ template<>
+        hash = c10::hash_combine(hash, N_hash);
+      }
+      // We don't need to hash the alias_info because it
+      // is not useful in distinguishing schema.
```
alias_info is tested in equality. Please also hash it.

wanchaol committed:
Update on "add __hash__ to FunctionSchema"


This PR adds __hash__ to FunctionSchema pybind binding, so that
it could be used for things like dict indexing

[ghstack-poisoned]

wanchaol committed:
Update on "add __hash__ to FunctionSchema"


This PR adds __hash__ to FunctionSchema pybind binding, so that
it could be used for things like dict indexing

[ghstack-poisoned]

ezyang commented on aten/src/ATen/core/alias_info.h:
```diff
@@ -117,3 +118,25 @@ inline std::ostream& operator<<(std::ostream& out, const AliasInfo& aliasInfo) {
+      auto hash = std::hash<bool>()(aliasInfo.isWrite());
+      for (auto &e: aliasInfo.beforeSets()) {
+        auto symbol_hash = std::hash<c10::Symbol>()(e);
+        hash = c10::hash_combine(hash, symbol_hash);
```
This is not correct. Iteration order over unordered set is not specified. Two unordered sets can have the same contents but differing iteration order. There are a few ways to solve this problem but a cheap and cheerful one is to use a commutative operator like addition or xor to combine the hashes from the operations in the set.

wanchaol replied:
Got it, thanks! Let me try this out.

wanchaol committed:
Update on "add __hash__ to FunctionSchema"


This PR adds __hash__ to FunctionSchema pybind binding, so that
it could be used for things like dict indexing

[ghstack-poisoned]

wanchaol commented:
@pytorchbot merge -f "test failures are lazy op test which is not related to the PR"
